### PR TITLE
access logging: gRPC logger retry to establish underlying stream connection

### PR DESCRIPTION
**Pull request by Shikugawa**
Signed-off-by: Shikugawa <rei@tetrate.io>

Commit Message: Add a retry mechanism to the grpc access logger. This retry mechanism currently only supports a simple retry count. Also, at the moment, retries are only fired when the gRPC stream fails to be established, and nothing happens if the stream is successfully established once and a reset is issued.
Additional Description:
Risk Level: Low
Testing: Unit
Docs Changes: Required
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: source/extensions/access_loggers/common/grpc_access_logger.h [hunk 1]
```diff
@@ -172,16 +191,18 @@ class GrpcAccessLogger : public Detail::GrpcAccessLogger<HttpLogProto, TcpLogPro
                    std::chrono::milliseconds buffer_flush_interval_msec,
                    uint64_t max_buffer_size_bytes, Event::Dispatcher& dispatcher,
                    Stats::Scope& scope, std::string access_log_prefix,
-                   const Protobuf::MethodDescriptor& service_method)
+                   const Protobuf::MethodDescriptor& service_method,
+                   const absl::optional<envoy::config::core::v3::RetryPolicy> retry_policy)
```
Comment (dmitri-d) [inline source/extensions/access_loggers/common/grpc_access_logger.h:195]: Here and in other spots:  several copies of retry_policy are being made -- perhaps pass it by reference or do a std::move?

File: source/extensions/access_loggers/common/grpc_access_logger.h [hunk 2]
```diff
@@ -135,10 +136,28 @@ template <typename LogRequest, typename LogResponse> class GrpcAccessLogClient {
...
 
+  Http::AsyncClient::StreamOptions createStreamOptionsForRetry() {
+    auto opt = Http::AsyncClient::StreamOptions();
+
+    if (!grpc_stream_retry_policy_) {
+      return opt;
+    }
+
+    envoy::config::route::v3::RetryPolicy retry_policy;
+    retry_policy.mutable_num_retries()->set_value(
```
Comment (dmitri-d) [inline source/extensions/access_loggers/common/grpc_access_logger.h:147]: I think that only `num_retries` from the retry policy is being used needs to be documented.

Comment (Shikugawa) [reply source/extensions/access_loggers/common/grpc_access_logger.h:147]: added

**Comment by dmitri-d** [comment]
lgtm!

File: source/extensions/access_loggers/common/grpc_access_logger.h [hunk 3]
```diff
@@ -77,13 +77,15 @@ template <typename GrpcAccessLogger, typename ConfigProto> class GrpcAccessLogge
 template <typename LogRequest, typename LogResponse> class GrpcAccessLogClient {
 public:
   GrpcAccessLogClient(const Grpc::RawAsyncClientSharedPtr& client,
-                      const Protobuf::MethodDescriptor& service_method)
-      : GrpcAccessLogClient(client, service_method, absl::nullopt) {}
+                      const Protobuf::MethodDescriptor& service_method,
+                      const absl::optional<envoy::config::core::v3::RetryPolicy> retry_policy)
```
Comment (lizan) [inline source/extensions/access_loggers/common/grpc_access_logger.h:81]: const ref?

File: source/extensions/access_loggers/common/grpc_access_logger.h [hunk 4]
```diff
@@ -77,13 +77,15 @@ template <typename GrpcAccessLogger, typename ConfigProto> class GrpcAccessLogge
...
+      : GrpcAccessLogClient(client, service_method, retry_policy, absl::nullopt) {}
   GrpcAccessLogClient(const Grpc::RawAsyncClientSharedPtr& client,
                       const Protobuf::MethodDescriptor& service_method,
+                      const absl::optional<envoy::config::core::v3::RetryPolicy> retry_policy,
```
Comment (lizan) [inline source/extensions/access_loggers/common/grpc_access_logger.h:85]: ditto

File: source/extensions/access_loggers/common/grpc_access_logger.h (hunk 2 above)
Comment (lizan) [reply source/extensions/access_loggers/common/grpc_access_logger.h:147]: Why only `num_retries` applies?

**Review by lizan — COMMENTED** [review]
/wait

File: source/extensions/access_loggers/common/grpc_access_logger.h (hunk 2 above)
Comment (Shikugawa) [reply source/extensions/access_loggers/common/grpc_access_logger.h:147]: I think that we don't need backoff strategy in our usecase. Our usecase is just to reconnect other cluster periodically, not to reconnect same cluster. So to speak, we don't expect that to overwhelm periodical pressure in current usecase, then backoff is not needed right now. This is why I add only `num_retries`.

Comment (Shikugawa) [reply source/extensions/access_loggers/common/grpc_access_logger.h:147]: My expectation to use this feature is to configure `num_retries` as the same number of cluster hosts.

Comment (Shikugawa) [reply source/extensions/access_loggers/common/grpc_access_logger.h:147]: added backoff

File: source/extensions/access_loggers/common/grpc_access_logger.h [hunk 5]
```diff
@@ -135,10 +136,40 @@ template <typename LogRequest, typename LogResponse> class GrpcAccessLogClient {
...
+      const auto base_interval_ms = PROTOBUF_GET_MS_REQUIRED(base_retry_backoff, base_interval);
+
+      auto* mutable_retry_back_off = retry_policy.mutable_retry_back_off();
+      mutable_retry_back_off->mutable_base_interval()->CopyFrom(
+          Protobuf::util::TimeUtil::MillisecondsToDuration(base_interval_ms));
+      mutable_retry_back_off->mutable_max_interval()->CopyFrom(
+          Protobuf::util::TimeUtil::MillisecondsToDuration(
+              PROTOBUF_GET_MS_OR_DEFAULT(base_retry_backoff, max_interval, 10 * base_interval_ms)));
+    }
+    retry_policy.set_retry_on("connect-failure");
```
Comment (lizan) [inline source/extensions/access_loggers/common/grpc_access_logger.h:161]: Why we need this part but not use `grpc_stream_retry_policy_` directly?

Comment (Shikugawa) [reply source/extensions/access_loggers/common/grpc_access_logger.h:161]: this is because we'd like to set `RetryBackOff` but provided type is `BackoffStrategy`.

Comment (Shikugawa) [reply source/extensions/access_loggers/common/grpc_access_logger.h:163]: add TODO comment to refactor here.

File: source/extensions/access_loggers/common/grpc_access_logger.h [hunk 6]
```diff
@@ -135,10 +136,42 @@ template <typename LogRequest, typename LogResponse> class GrpcAccessLogClient {
...
 
+  Http::AsyncClient::StreamOptions createStreamOptionsForRetry() {
+    auto opt = Http::AsyncClient::StreamOptions();
+
+    if (!grpc_stream_retry_policy_) {
+      return opt;
+    }
+
+    // TODO(shikugawa): refactor conversion from route retry policy to core retry policy.
+    // This conversion is duplicated with jwks_fetcher.cc.
```
Comment (lizan) [inline source/extensions/access_loggers/common/grpc_access_logger.h:147]: Lets do this in another PR before this lands.

**Review by lizan — CHANGES_REQUESTED** [review]

File: source/extensions/access_loggers/common/grpc_access_logger.h (hunk 6 above)
Comment (Shikugawa) [reply source/extensions/access_loggers/common/grpc_access_logger.h:147]: done https://github.com/envoyproxy/envoy/pull/17803

**Comment by Shikugawa** [comment]
@lizan fixed with `convertCoreToRoutePolicy()`. PTAL.

**Review by lizan — DISMISSED** [review]
@htuch can you do a non-tetrands review?

**Comment by htuch** [comment]
I don't think there is a solid story for how to get the retry/buffering semantics aligned across the gRPC implementations. This really touches on the fact that we have not made any progress on unifying the gRPC implementations more broadly.

I think for this PR, maybe just clarify in the release notes and comments that this only works with the Envoy gRPC client and we can merge.

**Comment by lizan** [comment]
@htuch ping?

**Review by htuch — APPROVED** [review]
LGTM, thanks!

**Comment by durd07** [comment]
@htuch @Shikugawa  the merge bring back the generated_api_shadow into main repo.
can we add a hook to check whether generated_api_shadow added by mistake?
I have seen serval times generated_api_shadow back.

**Comment by Shikugawa** [comment]
> @htuch @Shikugawa the merge bring back the generated_api_shadow into main repo. can we add a hook to check whether generated_api_shadow added by mistake? I have seen serval times generated_api_shadow back.

Thank you for the notification. This is no longer needed and just my mistake. https://github.com/envoyproxy/envoy/pull/18911

**Comment by htuch** [comment]
@durd07 what other examples are there? If this is common I can add a hook, otherwise I think we will soon forget about generated_api_shadow/ and this won't sneak back in.

**Comment by durd07** [comment]
@htuch , yes, you're right. maybe we will forget about it soon. I met this situation at least twice. but I think it is not necessary to emphasis the alreay not exist codes. please forget it.
